Architecture: Generic, template-able resource reconciler  (Espejote) (#368)
appuio/openshift4-docs: docs/modules/ROOT/pages/references/architecture/espejote-in-cluster-templating-controller.adoc

We got an extensive array of tools to manage these configurations, from bash "reconcilers," to cron jobs, to policy engines to fully fledged custom controllers.
Those tools are all specialized and don't cover all use-cases.

A maintainable, fully instrumented, and extensible controller that can manage arbitrary configurations in a cluster is missing.
We decided to try our hand at creating one.

== High Level Goals

* We can manage arbitrary configurations in a cluster
** The configuration may depend on external state
** We can apply partial configurations to externally managed resources
* The amount of requests and caching is tuneable
* We can monitor failures and successes and have a tight feedback loop

== Non-Goals

* Centralized configuration management or GitOps
* Policy enforcement
* Security or access control

== Name

"Espejote" for "big mirror" in Spanish.
It's a play on the earlier project "Espejo" which means "mirror."
"Espejo" was a tool to mirror resources to multiple namespaces.
"Espejote" is a more general tool to manage arbitrary configurations in a cluster.

== Implementation

The controller is based on `controller-runtime`.

> It's managed through a Custom Resource Definition (CRD) called `(Cluster)ManifestTemplate`.

Author: ```suggestion
It's managed through a Custom Resource Definition (CRD) called `(Cluster)ManagedResource`.
```

    {
      myFunction: function() 42,
    }
  hello-world.libsonnet: |
    {
      helloWorld: function(name) 'Hello, %s!' % [name],
    }
----

=== ManagedResource

[source,yaml]
----
apiVersion: espejote.io/v1alpha1
kind: ManagedResource
metadata:
  name: copy-configmap
  namespace: my-namespace <1>
spec:
  triggers:
  - interval: 10s
  - watchResource:
      apiVersion: v1
      kind: ConfigMap
      # name: my-configmap
      labelSelector:
        matchExpressions:
        - key: espejote.io/created-by
          operator: DoesNotExist
      minTimeBetweenTriggers: 10s
  context: <2>
  - def: cm <3>
    resource: <4>
      apiVersion: v1
      kind: ConfigMap
      # name: my-configmap <5>
      labelSelector: <6>
        matchExpressions:
        - key: espejote.io/created-by
          operator: DoesNotExist
      cache: true
  template: |
    local cm = std.extVar('context').cm; <7>

    [ <8>
      c {
        metadata: {
          name: 'copy-of-' + c.metadata.name,
          // namespace: c.metadata.namespace, <9>
          labels+: {
            'espejote.io/created-by': 'copy-configmap'
          }
        }
      },
      for c in cm.items <10>
    ]
----

> <1> `ManifestTemplate` are always namespace-scoped and can't access any resources outside of their namespace.

Author: ```suggestion
<1> `ManagedResource` are always namespace-scoped and can't access any resources outside of their namespace.
```

= Espejote: An in-cluster templating controller

== Problem Statement

Not every cluster configuration is fully static and can be managed through GitOps.

Some configurations depend on the cluster state or might want to be managed by upstream, by end-users, or by third-party products directly.

We got an extensive array of tools to manage these configurations, from bash "reconcilers," to cron jobs, to policy engines to fully fledged custom controllers.
Those tools are all specialized and don't cover all use-cases.

A maintainable, fully instrumented, and extensible controller that can manage arbitrary configurations in a cluster is missing.

> We decided to try our hand at creating one.

Reviewer: Maybe link to the decision here:

```suggestion
xref:oc4:ROOT:explanations/decisions/prometheusrule-controller.adoc[We decided] to try our hand at creating one.
```

== High Level Goals

* We can manage arbitrary configurations in a cluster

> ** The configuration may depend on external state

Reviewer: maybe clarify external state here? is the state external, but in-cluster? or external behind an arbitrary URL?

Reviewer 2: clarified as `state managed by other controllers inside the cluster`

<15> `std.extVar` is used to access the context.
It returns a object with all defined resources as keys.
<16> The template can return a list of resources or a single resource.
<17> The namespace of the resource will always be overwritten to the namespace of the `ManagedResource`.
<18> Resource variables are always a list with zero or more items.

=== `ClusterManagedResource`

[source,yaml]
----
apiVersion: espejote.io/v1alpha1
kind: ClusterManagedResource
metadata:
  name: default-network-policies
spec:
  triggers:
  - watchResource:
      apiVersion: v1
      kind: Namespace <1>
      labelSelector:
        matchExpressions:
        - key: inject-cm.syn.tools
          operator: Exists
  template: |
    local cm = std.extVar('trigger').resource; <2>

    [
      {
        metadata: {
          name: 'injected-cm' + c.metadata.name,
          namespace: c.metadata.namespace, <3>
        }
        stringData: {
          'my-key': 'my-value',
        }
      }
    ]
----
<1> Cluster scoped `ClusterManagedResource` can access all resources in the cluster.
This includes cluster scoped resources.
<2> The full manifest of the triggering is available.

> <3> A cluster managed resource can create resources in any namespace and thus needs to set the namespace.

Reviewer: this is a bit inconsistent/incomplete, and I'm not sure what the actual example wants to do. I'd like to have a consistent example here (either copy config map to target namespace -- in that case the source is missing, or generating netpols, in which case the full object should be visible in Jsonnet?

== Sample use-cases

=== Sync upgrade notifications with a ArgoCD sync hook and bash script

We want to show a notification in the OpenShift Console when a minor upgrade is available.
To implement this we added a job triggered by an ArgoCD sync.
The job is reading another resource and creating a console notification manifest from it.

* https://github.com/appuio/component-openshift4-console/blob/740628ebf3822ea82a64fade1e42eb9ff52f67c7/tests/golden/upgrade-notification/openshift4-console/openshift4-console/31_upgrade_notification.yaml#L63[component-openshift4-console ArgoCD sync hook]
* https://github.com/appuio/component-openshift4-console/blob/740628ebf3822ea82a64fade1e42eb9ff52f67c7/component/scripts/create-console-notification.sh[component-openshift4-console bash script]

=== Sync OpenShift Console TLS Secret with a bash "reconciler"

> Cert-manager only allows to create certificates in the same namespace as the ingress.

Reviewer: ```suggestion
Cert-manager only allows to create the secret holding the certificate data  in the same namespace as the `Certificate` resource.
```

> OpenShift Console ingress is in the `openshift-console` namespace, but the secret needs to be applied in the `openshift-config` namespace.

Reviewer: This isn't directly coupled with the OpenShift console ~~ingress~~ route

```suggestion
The OpenShift Console route is in the `openshift-console` namespace, but the secret needs to be applied in the `openshift-config` namespace.
However, we must create the `Certificate` resource in `openshift-console`, since otherwise the OpenShift ingress doesn't admit the HTTP challenge ingress.
```

The full manifest is available if using the `watchResource` trigger.
<15> `std.extVar` is used to access the context.
It returns a object with all defined resources as keys.
<16> The template can return a list of resources or a single resource.
<17> The namespace of the resource will always be overwritten to the namespace of the `ManagedResource`.
<18> Resource variables are always a list with zero or more items.

=== `ClusterManagedResource`

[source,yaml]
----
apiVersion: espejote.io/v1alpha1
kind: ClusterManagedResource
metadata:
  name: inject-configmaps
spec:
  triggers:
  - watchResource:
      apiVersion: v1
      kind: Namespace <1>
      labelSelector:
        matchExpressions:
        - key: inject-cm.syn.tools
          operator: Exists
  context:
  - def: base
    resource:
      apiVersion: v1
      kind: ConfigMap
      name: cm-to-inject
  template: |
    local ns = std.extVar('trigger').resource; <2>
    local base = std.extVar('context').base[0];

    [
      {
        apiVersion: 'v1',
        kind: 'ConfigMap',
        metadata: {
          name: 'injected-cm',
          namespace: ns.metadata.name, <3>
        }
        data: base.data,
      }
    ]
----

> <1> Cluster scoped `ClusterManagedResource` can access all resources in the cluster.
> This includes cluster scoped resources.

Reviewer: ```suggestion
<1> Cluster scoped `ClusterManagedResource` can access all resources in the cluster.
This includes cluster scoped resources such as `Namespace`.
```

== Implementation

The controller is based on `controller-runtime`.
It's managed through a Custom Resource Definition (CRD) called `(Cluster)ManagedResource`.
The controller uses Jsonnet to render manifests.
Any Kubernetes manifest can be added to the templating context.

Reconcile triggers are separated from the context resources to allow for more fine-grained control over the reconciliation.

=== Reconcile triggers

Reconcile triggers are used to re-render the template.
They're separated from the context resources to allow for more fine-grained control.

[source,yaml]
----
triggers:
- interval: 10s
- watchResource:
    apiVersion: v1
    kind: ConfigMap
    labelSelector:
      matchExpressions:
      - key: espejote.io/created-by
        operator: DoesNotExist
    minTimeBetweenTriggers: 10s
- watchResource:
    apiVersion: v1
    kind: ConfigMap
    name: my-configmap
----

The trigger that fires is exposed to the Jsonnet as the `trigger` variable.

[source,jsonnet]
----
local trigger = std.extVar('trigger');
----

Currently, we support two types of triggers:

==== `interval`

The template is re-rendered every `interval`.

==== `watchResource`

> The template is re-rendered whenever the watched resource is created, changed, or deleted.

Reviewer 2: That's a good idea! `watchResource` is the wrong context because it's input only and we do not really know what the underlying event was.

I'll add `applyOptions.createOnly` to use for only creating the resulting resource from the template.
